File: android/app/src/main/java/com/commuteoptimizer/widget/data/api/GoogleRoutesService.kt
package com.commuteoptimizer.widget.data.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Service to call Google Directions API directly for transit routes.
 * Returns full transit routes with all transfer details.
 */
object GoogleRoutesService {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    data class TransitStep(
        val line: String,
        val vehicle: String?,
        val departureStop: String?,
        val arrivalStop: String?,
        val numStops: Int?,
        val duration: Int?
    )

    data class RouteResult(
        val status: String,
        val durationMinutes: Int?,
        val distance: String?,
        val transitSteps: List<TransitStep>
    )

    suspend fun getTransitRoute(
        apiKey: String,
        originLat: Double,
        originLng: Double,
        destLat: Double,
        destLng: Double
    ): RouteResult = withContext(Dispatchers.IO) {
        val url = "https://maps.googleapis.com/maps/api/directions/json" +
            "?origin=$originLat,$originLng" +
            "&destination=$destLat,$destLng" +
            "&mode=transit" +
            "&departure_time=now" +
            "&key=$apiKey"

        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        try {
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return@withContext errorResult("No response")

            val json = JSONObject(body)
            val status = json.optString("status", "ERROR")

            if (status != "OK") {
                return@withContext RouteResult(status, null, null, emptyList())
            }

            val routes = json.optJSONArray("routes")
            if (routes == null || routes.length() == 0) {
                return@withContext RouteResult("NO_ROUTES", null, null, emptyList())
            }

            val route = routes.getJSONObject(0)
            val legs = route.optJSONArray("legs")
            if (legs == null || legs.length() == 0) {
                return@withContext RouteResult("NO_LEGS", null, null, emptyList())
            }

            val leg = legs.getJSONObject(0)
            val durationSeconds = leg.optJSONObject("duration")?.optInt("value", 0) ?: 0
            val durationMinutes = durationSeconds / 60
            val distance = leg.optJSONObject("distance")?.optString("text", null)

            val steps = leg.optJSONArray("steps") ?: return@withContext RouteResult(
                status, durationMinutes, distance, emptyList()
            )

            val transitSteps = mutableListOf<TransitStep>()
            for (i in 0 until steps.length()) {
                val step = steps.getJSONObject(i)
                val travelMode = step.optString("travel_mode", "")

                if (travelMode == "TRANSIT") {
                    val transitDetails = step.optJSONObject("transit_details") ?: continue
                    val line = transitDetails.optJSONObject("line")
                    val shortName = line?.optString("short_name", null)
                        ?: line?.optString("name", "?")
                        ?: "?"

                    val vehicle = line?.optJSONObject("vehicle")?.optString("type", null)
                    val departureStop = transitDetails.optJSONObject("departure_stop")?.optString("name", null)
                    val arrivalStop = transitDetails.optJSONObject("arrival_stop")?.optString("name", null)
                    val numStops = if (transitDetails.has("num_stops")) transitDetails.optInt("num_stops") else null
                    val stepDuration = step.optJSONObject("duration")?.optInt("value", 0)?.let { it / 60 }

                    transitSteps.add(TransitStep(
                        line = cleanLineName(shortName),
                        vehicle = vehicle,
                        departureStop = departureStop,
                        arrivalStop = arrivalStop,
                        numStops = numStops,
                        duration = stepDuration
                    ))
                }
            }

            RouteResult(status, durationMinutes, distance, transitSteps)
        } catch (e: Exception) {
            errorResult(e.message ?: "Unknown error")
        }
    }

    private fun cleanLineName(name: String): String {
        return name
            .replace(" Line", "")
            .replace(" Train", "")
            .replace("Exp", "")
            .trim()
    }

    private fun errorResult(error: String) = RouteResult("ERROR", null, null, emptyList())
}

File: ios/CommuteOptimizer/CommuteOptimizer/Services/GoogleRoutesService.swift
import Foundation

/// Service to call Google Directions API directly for transit routes.
/// Returns full transit routes with all transfer details.
/// Matches Android implementation for feature parity.
actor GoogleRoutesService {
    struct TransitStep {
        let line: String
        let vehicle: String?
        let departureStop: String?
        let arrivalStop: String?
        let numStops: Int?
        let duration: Int?
    }

    struct RouteResult {
        let status: String
        let durationMinutes: Int?
        let distance: String?
        let transitSteps: [TransitStep]

        static let error = RouteResult(status: "ERROR", durationMinutes: nil, distance: nil, transitSteps: [])
    }

    func getTransitRoute(
        apiKey: String,
        originLat: Double, originLng: Double,
        destLat: Double, destLng: Double
    ) async -> RouteResult {
        guard !apiKey.isEmpty else {
            return .error
        }

        let urlString = "https://maps.googleapis.com/maps/api/directions/json" +
            "?origin=\(originLat),\(originLng)" +
            "&destination=\(destLat),\(destLng)" +
            "&mode=transit" +
            "&departure_time=now" +
            "&key=\(apiKey)"

        guard let url = URL(string: urlString) else {
            return .error
        }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let status = json["status"] as? String else {
                return .error
            }

            if status != "OK" {
                return RouteResult(status: status, durationMinutes: nil, distance: nil, transitSteps: [])
            }

            guard let routes = json["routes"] as? [[String: Any]],
                  let route = routes.first,
                  let legs = route["legs"] as? [[String: Any]],
                  let leg = legs.first else {
                return RouteResult(status: "NO_ROUTES", durationMinutes: nil, distance: nil, transitSteps: [])
            }

            let durationSeconds = (leg["duration"] as? [String: Any])?["value"] as? Int ?? 0
            let durationMinutes = durationSeconds / 60
            let distance = (leg["distance"] as? [String: Any])?["text"] as? String

            guard let steps = leg["steps"] as? [[String: Any]] else {
                return RouteResult(status: status, durationMinutes: durationMinutes, distance: distance, transitSteps: [])
            }

            var transitSteps: [TransitStep] = []
            for step in steps {
                let travelMode = step["travel_mode"] as? String ?? ""

                if travelMode == "TRANSIT" {
                    guard let transitDetails = step["transit_details"] as? [String: Any] else { continue }

                    let line = transitDetails["line"] as? [String: Any]
                    let shortName = (line?["short_name"] as? String) ?? (line?["name"] as? String) ?? "?"

                    let vehicle = (line?["vehicle"] as? [String: Any])?["type"] as? String
                    let departureStop = (transitDetails["departure_stop"] as? [String: Any])?["name"] as? String
                    let arrivalStop = (transitDetails["arrival_stop"] as? [String: Any])?["name"] as? String
                    let numStops = transitDetails["num_stops"] as? Int
                    let stepDuration = ((step["duration"] as? [String: Any])?["value"] as? Int).map { $0 / 60 }

                    transitSteps.append(TransitStep(
                        line: cleanLineName(shortName),
                        vehicle: vehicle,
                        departureStop: departureStop,
                        arrivalStop: arrivalStop,
                        numStops: numStops,
                        duration: stepDuration
                    ))
                }
            }

            return RouteResult(
                status: status,
                durationMinutes: durationMinutes,
                distance: distance,
                transitSteps: transitSteps
            )
        } catch {
            return .error
        }
    }

    private func cleanLineName(_ name: String) -> String {
        var cleaned = name
            .replacingOccurrences(of: " Line", with: "")
            .replacingOccurrences(of: " Train", with: "")
            .replacingOccurrences(of: "Exp", with: "")
            .trimmingCharacters(in: .whitespaces)
            .uppercased()

        // Express variants: 6X -> 6, 7X -> 7, FX -> F
        if cleaned.count == 2 && cleaned.hasSuffix("X") {
            cleaned = String(cleaned.dropLast())
        }

        return cleaned
    }
}
